### PR TITLE
Privacy notification controls

### PR DESCRIPTION
**Pull request by DominicHolmes**
Created 2 screens -- Notifications and Privacy. Both are on the more page, below the edit profile VC.

- `UserDefaults.shared.getNotificationPreferences()` retrieves a `NotificationPreferences` object, which is just a dictionary of <NotificationOption : Bool>. NotificationOption is an enum with values for the different kinds of notifications. Each enum option also has associated human readable string values.
- The same is done for PrivacyPreferences.
- Nothing is at this time uploaded to our server -- everything stored in UserDefaults.

IMG_1003
IMG_1002

### CODE REVIEW

**Comment by DominicHolmes** [comment]
Before merging this to development, we should make sure
1. we respect the preferences being toggled
2. we upload preferences to the server, and make sure they're being honored there (notifications)

File: PennMobile/General/UserDefaults + Helpers.swift [hunk 1]
```diff
@@ -389,3 +391,47 @@ extension UserDefaults {
...
+    }
+}
+
+// MARK: - Notification Settings
+extension UserDefaults {
+    func save(_ preferences: NotificationPreferences) {
+        let encoder = JSONEncoder()
+        if let encoded = try? encoder.encode(preferences) {
+            UserDefaults.standard.set(encoded, forKey: UserDefaultsKeys.notificationPreferences.rawValue)
+        }
```
Comment (joshdoman) [inline PennMobile/General/UserDefaults + Helpers.swift]: Missing `synchronized()`

File: PennMobile/General/UserDefaults + Helpers.swift [hunk 2]
```diff
@@ -389,3 +391,47 @@ extension UserDefaults {
...
     }
 }
+
+// MARK: - Privacy Settings
+extension UserDefaults {
+    func save(_ preferences: PrivacyPreferences) {
+        let encoder = JSONEncoder()
+        if let encoded = try? encoder.encode(preferences) {
+            UserDefaults.standard.set(encoded, forKey: UserDefaultsKeys.privacyPreferences.rawValue)
+        }
```
Comment (joshdoman) [inline PennMobile/General/UserDefaults + Helpers.swift]: Missing `synchronized`

File: PennMobile/Notifications/NotificationPreference.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//
+//  Created by Dominic Holmes on 12/27/19.
+//  Copyright © 2019 PennLabs. All rights reserved.
+//
+
+import Foundation
+
+typealias NotificationPreferences = Dictionary<NotificationOption, Bool>
+
+enum NotificationOption: String, Codable, CaseIterable {
```
Comment (joshdoman) [inline PennMobile/Notifications/NotificationPreference.swift]: Are you using `CaseIterable` anywhere? It doesn't seem so. If not, we should delete it.

File: PennMobile/About/Privacy/PrivacyPreference.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+//
+//  Created by Dominic Holmes on 12/27/19.
+//  Copyright © 2019 PennLabs. All rights reserved.
+//
+
+import Foundation
+
+typealias PrivacyPreferences = Dictionary<PrivacyOption, Bool>
+
+enum PrivacyOption: String, Codable, CaseIterable {
```
Comment (joshdoman) [inline PennMobile/About/Privacy/PrivacyPreference.swift]: Are you using CaseIterable anywhere? It doesn't seem so. If not, we should delete it.

File: PennMobile/Notifications/NotificationsTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        label.widthAnchor.constraint(equalTo: widthAnchor, constant: -90.0).isActive = true
+        label.topAnchor.constraint(equalTo: topAnchor, constant: 7.0).isActive = true
+        label.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -7.0).isActive = true
+        
+        label.font = UIFont.systemFont(ofSize: 17, weight: .medium)
+        label.numberOfLines = 0
+        label.minimumScaleFactor = 0.7
+        label.adjustsFontSizeToFitWidth = true
+        
+        label.text = " "
```
Comment (joshdoman) [inline PennMobile/Notifications/NotificationsTableViewCell.swift]: Delete this line.

File: PennMobile/About/Privacy/PrivacyTableViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        label.widthAnchor.constraint(equalTo: widthAnchor, constant: -90.0).isActive = true
+        label.topAnchor.constraint(equalTo: topAnchor, constant: 7.0).isActive = true
+        label.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -7.0).isActive = true
+        
+        label.font = UIFont.systemFont(ofSize: 17, weight: .medium)
+        label.numberOfLines = 0
+        label.minimumScaleFactor = 0.7
+        label.adjustsFontSizeToFitWidth = true
+        
+        label.text = " "
```
Comment (joshdoman) [inline PennMobile/About/Privacy/PrivacyTableViewCell.swift]: Delete this line

File: PennMobile/General/UserDefaults + Helpers.swift (hunk 1 above)
Comment (DominicHolmes) [reply PennMobile/General/UserDefaults + Helpers.swift]: resolved -- this function no longer needed

**Comment by DominicHolmes** [comment]
@joshdoman made your changes. Waiting to merge this until we have all the server stuff finished.

File: PennMobile/General/UserDefaults + Helpers.swift [hunk 7]
```diff
@@ -389,3 +391,78 @@ extension UserDefaults {
...
+    func set(_ notificationOption: NotificationOption, to newValue: Bool) {
+        var prefs = getAllNotificationPreferences()
+        prefs[notificationOption.rawValue] = newValue
+        saveAll(notificationPreferences: prefs)
+    }
+    
+    // Get values for each notification option (default to false if no preference exists)
+    func getPreference(for option: NotificationOption) -> Bool {
+        let prefs = getAllNotificationPreferences()
+        return prefs[option.rawValue] ?? false
```
Comment (joshdoman) [inline PennMobile/General/UserDefaults + Helpers.swift]: I think Notification preferences should be opt-out, not opt-in, so I think the default notification preference should be true. Most apps I know are like that, and there's very little cost to the user to receive a GSR reminder. Plus, it keeps the "authorize push notification" flow simple: just yes or no. Do you agree?

File: PennMobile/General/Networking + Analytics/UserDBManager.swift [hunk 8]
```diff
@@ -243,6 +243,71 @@ extension UserDBManager {
...
+        
+        struct CodableUserSettings: Codable {
+            let notifications: NotificationPreferences
+            let privacy: PrivacyPreferences
+        }
+        
+        OAuth2NetworkManager.instance.getAccessToken { (token) in
+            guard let token = token else {
+                // User has not logged into Platform
+                return
```
[the remark this answers is not in the document]
Comment (DominicHolmes) [reply PennMobile/General/Networking + Analytics/UserDBManager.swift]: resolved in updated code

File: PennMobile/General/UserDefaults + Helpers.swift (hunk 7 above)
Comment (DominicHolmes) [reply PennMobile/General/UserDefaults + Helpers.swift]: Made the default true -- great suggestion! I totally agree.

**Comment by DominicHolmes** [comment]
Great improvements! Totally forgot about the notifications enabled thing. I go back and forth about the 1sec delay on settings changes. But it does make it feel more substantial so I'm cool with it.

Are we set to merge this? @joshdoman

**Pull request merged**